backend/upload/models.py

```diff
@@ -1,3 +1,331 @@
+                        # Only add to values set if not too large -
+                        # we only use this to test if a filter should be
+                        # created
+                        m['values'].add(value)
+                    if m.get('allnumeric', None):
+                        if value.isnumeric():
+                            m['min_value'] = min(m['min_value'], int(value))
+                            m['max_value'] = max(m['max_value'], int(value))
+                        else:
+                            m['allnumeric'] = False
+                else:
+                    m = {}
+                    m['type'] = type_
+                    m['values'] = {value}
+                    if value.isnumeric():
+                        m['allnumeric'] = True
+                        m['min_value'] = int(value)
+                        m['max_value'] = int(value)
+                    self._metadata[name] = m
+
+    def _probe_file(self, path):
+        '''Probe file format to allow autodiscovery'''
+        filename = str(path)
+        if filename.lower().endswith('.txt'):
+            return self.TXT
+        elif filename.lower().endswith('.cha'):
+            return self.CHAT
+        elif filename.lower().endswith('.xml'):
+            with open(filename, 'r') as f:
+                firstline = f.readline()
+                secondline = f.readline()
+                if firstline.startswith('<FoLiA'):
+                    return self.FOLIA
+                elif secondline.startswith('<alpino_ds'):
+                    return self.ALPINO
+        return None
+
+    def _unpack(self):
+        '''Unpack compressed input file and set input_dir.
+        TODO: not yet implemented.'''
+        if not self.input_file:
+            raise UploadError('Need input_file to unpack')
+        raise UploadError('Unpacking not yet implemented')
+
+    def _read_input_files(self):
+        '''Divide extracted files into components and probe input format'''
+        inputpath = Path(self.input_dir)
+
+        # Put all toplevel files in a 'main' component and all files
+        # in a subdirectory in a component with the name of the directory
+        components = {}
+        for entry in inputpath.glob('*'):
+            if entry.is_file():
+                format = self._probe_file(entry)
+                if format:
+                    if self.input_format and format != self.input_format:
+                        raise UploadError(
+                            'Different input formats found ({} and {}).'
+                            .format(self.input_format, format)
+                        )
+                    self.input_format = format
+                    if 'main' not in components:
+                        # TODO what to do if one dir is named main?
+                        components['main'] = []
+                    components['main'].append(entry)
+            if entry.is_dir():
+                files = entry.glob('**/*')
```

Comment:
maybe these two if clauses could be combined to one glob? or at least the body blocks could be refactored into a method